design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -0,0 +1,442 @@
+**Release Target**: TBD |
+**Status**: _In Discussion_ |
+**Discussion**: TBD
+
+**tl;dr:** _New tool to provide ksqlDB users for easy and automated schema migrations for their
+           ksqlDB environments. This allows users to version control their ksqlDB schema; recreate
+           their schema from scratch; and migrate their current schema to newer versions._
+
+## Motivation and background
+
+Schema migrations (also database migrations) refers to the process of performing updates or rollbacks on a database schema to a
+newer or older schema version. This practice is pretty common in any application lifecycle. Users use a version control for their
+applications which allow them to know if an application requires an upgrade or rolling back a buggy change. The same is necessary
+with a database schema. Users look for ways to version control schema changes along the application lifecycle.
+
+Existing tools exist to perform schema migrations on any database (MySQL, Postgres, Oracle, etc). These tools allow users to
+version the database schema, then automate the process to easily upgrade the schema to a newer version. If a schema is buggy,
+then users can rollback to the previous schema version. One more reason for using these tools is the integration with any
+CI/CD system, which allows users to test and verify a schema upgrade will work as expected.
+
+You can learn more about these existing tools:
+- [Flyway](https://flywaydb.org/)
+- [Liquibase](https://www.liquibase.org/)
+- [DBGeni](http://dbgeni.appsintheopen.com/index.html)
+
+ksqlDB requires the same kind of integration with CI/CD systems and automation to deploy ksqlDB schema changes in any environment.
+Some tools, like `Flyway`, support plugins to integrate schema migrations with a different non-supported database. However, our
+syntax is exclusively to ksqlDB; tools require JDBC support; and tools have several features that we cannot support
+(i.e. transactions & databases); so this integration becomes complex or unsupported. For that reason, ksqlDB must have its own
+schema migrations tool to provide the same migrations benefits to users who want to easily deploy and automate ksqlDB schema
+changes in all their ksqlDB environments.
+
+These benefits include:
+- Version the ksqlDB schema environments
+- Integrate ksqlDB schema changes with CI/CD environments
+- Simplify schema evolution changes across different environments
+
+This KLIP proposes a new tool for schema migrations. You will learn about the design aspects of the tool, and the features
+to support for a basic schema migration process.
+
+## What is in scope
+
+* Discuss design details for a new tool that provides schema migrations support for ksqlDB
+
+    Basic features to support:
+    - New CLI and API that can easily integrate with CI/CD environments
+    - Apply migrations on any ksqlDB environments
+    - Rollback migrated schemas to previous versions
+    - Version control ksqlDB schema
```

Comment:
One other thing that I think would be really useful would be producing a consolidated view of all the statements in one file. Rails, and I think Django, do this so you can see what is going on in one place. This consolidated view is typically checked into version control, and is always treated as read-only. @derekjn had some ideas about how this relates to or could be subsumed by supporting schema dumping.

Reply:
Is this similar to Flyway's dry-runs feature? https://flywaydb.org/documentation/concepts/dryruns

Reply:
I talked with @derekjn about this. This is the functionality we expect to have when working on the dump/export command for ksqlDB, which will export all metadata as a SQL file.